### ISSUE TITLE
Can I drop the multiprocessing reactor into regular Klipper?

### ISSUE THREAD

**Issue by D4SK** [mfp20/klipper-ng#16]
I'm currently developing a native python UI plugin for Klipper, it currently runs in the Klipper process enabling responsive UX and rich communication with Klipper.

The one problem being that the additional CPU usage impedes with the real-time tasks of Klipper, the GIL prevents true parallelism. In my understanding your implementation of the reactor could solve these issues.

How compatible is it with Klipper, can I just drop it in only readjusting the naming changes?
How well does the current version work, do parallel operations not create inconsistent states in something like the kinematic position?

Thanks for any help!

**Comment by D4SK** [comment]
Still working on this, a quick comment would be appreciated.

**Comment by mfp20** [comment]
Apologize for the delay, I haven't seen this issue until now. 

**TLTR: In theory yes, my implementation works around the GIL. That's the main reason to re-implement the reactor. But my implementation isn't working at all.**

It does programmatically the same things the old one did but It has never been connected to a real mcu. The communication with mcu part is missing because I'm rewriting the mcu firmware; I need a general purpose firmware (ie: an abstraction that can be used in different applications, not only 3d printing) and I'd like to change the protocol in order to have 1 piece of code both as a firmware and as a library for the python part on host; currently instead, Klipper have a monolitic firmware, and a chelper for python. The original firmware, given the published performance, is incredibly efficient but the code is very unfriendly with new developers so I was hoping to solve this issue as well.

I've been stuck on rewriting the firmware in the past 10 months; attempted 10-12 different approaches, all failing badly. Recently I've scrapped the original firmware to the bare minimum, and [refactor to embed the code in this project](https://github.com/mfp20/klipper-firmware-ng). But then I stopped because I've found what looks like the abstraction I've been searching for: [avrshell](https://github.com/rahra/avrshell). So I'm working on it at the moment.
What avrshell does is... reading/writing memory locations instead of commanding steppers/thermocouples/etc. I've ported some of the smart features of klipper in there (ex: VLQ integers encoding, dynpool, memory barrier for avr) and replaced others (ex: COBS instead of HDLC). I didn't publish this code yet, but there's no much to publish as it barely compiles.

Anyway, I'd be happy if someone would like to fork the python code and keep working together on renovating Klipper. If our needs match, we could keep going together. Currently, as I said, I'm not working on the python part, so you can change anything you want without the need to cope with me; just don't trash my needs (ie: make my complex custom printer work!).
And the final goal, once being in alpha state (bare minimum in working state), is to ask a review to Kevin O'Connor from Klipper project; in case he would like to adopt this as v2.0. I mean, not to spawn an independent project.

**Comment by D4SK** [comment]
Ok thanks, will probably try to work with the multiprocessing reactor then. I'm using stm32 mcu's though.

**Comment by mfp20** [comment]
> Ok thanks, will probably try to work with the multiprocessing reactor then. I'm using stm32 mcu's though.

Well, the reactor by itself is a python 'class'. When I wrote 'remove klipper's unfriendliness to new developers' I meant exactly the chance to easily reuse the code. I hope it will be useful to you.

About stm32 mcu: I'm working to keep the firmware multiplatform as I have multiple mcus as well. I pointed you to avrshell as a concept only; I don't know yet whether to rewrite the assembly in C or write assembly code for each supported mcu, mostly because I don't know what tools are available for other mcus (SAMD, LPC, LX, STM32... never seen before).

But I don't have any stm32 at the moment. You might be the chance for me to test the new firmware... as I wrote before: if you are willing to support a bigger effort. The way to do it might be for you to keep in mind that the communication part of your python interface might change in the near future (ie: being able to communicate with my firmware).
Basically you should introduce some temporary overhead in your code:

pyif -> pyif_comms -> k_chelper -> klipper_fw

and once I finish my firmware would become:

pyif -> pyif_comms -> k_ng_fw
